### PR TITLE
Threshold index form

### PR DESCRIPTION
**Pull request by nhsiehgit**
Implements the threshold list view with editable in-line form rows

https://github.com/getsentry/sentry/assets/6186377/5b3c2f09-73ab-45d7-915b-ce30c9de0d5c

### CODE REVIEW

File: src/sentry/api/endpoints/release_thresholds/release_threshold_details.py [hunk 1]
```diff
@@ -27,8 +27,8 @@
 class ReleaseThresholdPUTSerializer(serializers.Serializer):
     threshold_type = serializers.ChoiceField(choices=ReleaseThresholdType.as_str_choices())
     trigger_type = serializers.ChoiceField(choices=ReleaseThresholdTriggerType.as_str_choices())
-    value = serializers.IntegerField()
-    window_in_seconds = serializers.IntegerField(min_value=0)
+    value = serializers.IntegerField(required=True, min_value=0)
+    serializers.IntegerField(required=True, min_value=0)
```
Comment (schew2381) [inline src/sentry/api/endpoints/release_thresholds/release_threshold_details.py:31]: I think the field got deleted by accident
```suggestion
    window_in_seconds = serializers.IntegerField(required=True, min_value=0)
```

File: src/sentry/api/endpoints/release_thresholds/release_threshold_details.py [hunk 2]
```diff
@@ -43,6 +43,7 @@ def validate_trigger_type(self, trigger_type: str):
 
 @region_silo_endpoint
 class ReleaseThresholdDetailsEndpoint(ProjectEndpoint):
+    permission_classes = (ProjectAlertRulePermission,)
```
Comment (schew2381) [inline src/sentry/api/endpoints/release_thresholds/release_threshold_details.py:46]: I checked this permission and it includes `alerts:read/write` but not `project:releases`. Does that make sense as I thought this would mirror the existing release endpoint permissions

File: src/sentry/api/endpoints/release_thresholds/release_threshold_details.py (hunk 1 above)
Comment (nhsiehgit) [reply src/sentry/api/endpoints/release_thresholds/release_threshold_details.py:31]: Ah oops thank you

File: src/sentry/api/endpoints/release_thresholds/release_threshold_details.py (hunk 2 above)
Comment (nhsiehgit) [reply src/sentry/api/endpoints/release_thresholds/release_threshold_details.py:46]: Ah thank you
yes - i'll switch to that

File: static/app/components/organizations/pageFilters/container.spec.tsx
Comment (evanpurkhiser) [inline static/app/components/organizations/pageFilters/container.spec.tsx:1]: Let's move all these page filter changes out into their own PR I think

File: static/app/utils/formatters.tsx [hunk 3]
```diff
@@ -181,7 +181,12 @@ export function getDuration(
  *
  * e.g. 1 hour 25 minutes 15 seconds
  */
-export function getExactDuration(seconds: number, abbreviation: boolean = false) {
+export function getExactDuration(
```
Comment (evanpurkhiser) [inline static/app/utils/formatters.tsx:184]: Let's move this out into it's own PR as well, I've got some feedback for this one as well I think

File: static/app/components/organizations/pageFilters/container.spec.tsx
Comment (nhsiehgit) [reply static/app/components/organizations/pageFilters/container.spec.tsx:1]: Oh oof thanks
I don't think i even need this

File: static/app/utils/formatters.tsx [hunk 4]
```diff
@@ -243,6 +258,63 @@ export function getExactDuration(seconds: number, abbreviation: boolean = false)
...
+ *
+ * @param value
+ * @param suffix
+ */
+export function deriveSeconds(value: number, suffix: string = 'seconds'): number {
+  switch (suffix) {
+    case 'minutes':
+      return value * SEC_IN_MIN;
+    case 'hours':
+      return value * SEC_IN_HR;
```
Comment (scttcper) [inline static/app/utils/formatters.tsx:310]: this is pretty simple but we do already have all of momentjs w/ duration support https://momentjs.com/docs/#/durations/ 

```ts
moment.duration(2, 'hours').as('seconds')
```

Comment (nhsiehgit) [reply static/app/utils/formatters.tsx:310]: Ah nice. This is exactly what i need.
will switch

File: static/app/views/releases/thresholdsList/thresholdGroupRows.tsx [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+    updatedEditingThresholds[thresholdId] = {
+      ...threshold,
+      windowValue,
+      windowSuffix,
+      hasError: false,
+    };
+    setEditingThresholds(updatedEditingThresholds);
+  };
+
+  const saveThreshold = thresholdIds => {
```
Comment (scttcper) [inline static/app/views/releases/thresholdsList/thresholdGroupRows.tsx:108]: types on these
```suggestion
  const saveThreshold = (thresholdIds: string[]) => {
```

File: static/app/views/releases/thresholdsList/thresholdGroupRows.tsx [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+    thresholdIds.forEach(id => {
+      const thresholdData = editingThresholds[id];
+      const submitData: {[key: string]: any} = {...thresholdData};
```
Comment (scttcper) [inline static/app/views/releases/thresholdsList/thresholdGroupRows.tsx:111]: maybe something like this would avoid casting it to any type
```suggestion
      const submitData= {
      	...thresholdData,
      window_in_seconds: seconds,
      environment: thresholdData.environment.name; // api expects environment as a string
      };
```

File: static/app/views/releases/thresholdsList/thresholdGroupRows.tsx [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+        .as('seconds');
+      submitData.window_in_seconds = seconds;
+      submitData.environment = thresholdData.environment.name; // api expects environment as a string
+      const request = api.requestPromise(path, {
+        method,
+        data: submitData,
+      });
+      request
+        .then(() => {
+          refetch();
```
Comment (scttcper) [inline static/app/views/releases/thresholdsList/thresholdGroupRows.tsx:130]: refetch might override things if they've made additional edits, just something to think about

File: static/app/views/releases/thresholdsList/thresholdGroupRows.tsx [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+        .catch(_err => {
+          setError('Issue saving threshold');
+          const errorThreshold = {
+            ...submitData,
+            hasError: true,
+            environment: thresholdData.environment, // convert local state environment back to object
+          };
+          const updatedEditingThresholds = {...editingThresholds};
+          updatedEditingThresholds[id] = errorThreshold as EditingThreshold;
+          setEditingThresholds(updatedEditingThresholds);
```
Comment (scttcper) [inline static/app/views/releases/thresholdsList/thresholdGroupRows.tsx:142]: this might need to use the current state since editingThresholds could be out of date when the promise finally resolves
```suggestion
          setEditingThresholds(editingThresholds => {
                    const errorThreshold = {
				            ...submitData,
				            hasError: true,
				            environment: thresholdData.environment, // convert local state environment back to object
				          };
				          const updatedEditingThresholds = {...editingThresholds};
				          updatedEditingThresholds[id] = errorThreshold as EditingThreshold;
         });
```

File: static/app/views/releases/thresholdsList/thresholdGroupRows.tsx [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+                <FlexCenter>
+                  <Input
+                    style={{width: '50%'}}
+                    value={threshold.windowValue}
+                    type="number"
+                    min={0}
+                    onChange={e =>
+                      editThresholdState(threshold.id, 'windowValue', e.target.value)
+                    }
+                  />
```
Comment (scttcper) [inline static/app/views/releases/thresholdsList/thresholdGroupRows.tsx:225]: putting the form into its own component with the edit state inside of it would cause fewer state changes on the entire list especially since i don't think you save more than 1 threshold at a time

File: static/app/views/releases/thresholdsList/index.spec.tsx [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    const query = {
+      project: [-1],
+    };
+    render(<ThresholdsList />, {context: routerContext, organization});
+    await waitFor(() =>
+      expect(mockThresholdFetch).toHaveBeenCalledWith(
+        '/organizations/test-thresholds/release-thresholds/',
+        expect.objectContaining({query})
+      )
+    );
```
Comment (scttcper) [inline static/app/views/releases/thresholdsList/index.spec.tsx:63]: wait for something from the request to be rendered from the component which helps make sure the component doesn't error on render and lets you remove the waitFor

```suggestion
			expect(await screen.findByText('something').toBeInTheDocument();
      expect(mockThresholdFetch).toHaveBeenCalledWith(
        '/organizations/test-thresholds/release-thresholds/',
        expect.objectContaining({query})
      )
```

File: static/app/views/releases/thresholdsList/thresholdGroupRows.tsx (hunk 8 above)
Comment (scttcper) [reply static/app/views/releases/thresholdsList/thresholdGroupRows.tsx:142]: i'm not sure why github is using tabs today 🤷

File: static/app/views/releases/thresholdsList/thresholdGroupRows.tsx (hunk 7 above)
Comment (nhsiehgit) [reply static/app/views/releases/thresholdsList/thresholdGroupRows.tsx:130]: Yup, already considered that!
Since the "in flight edits" are already set in state - their form data actually won't be affected by the refetch :)

File: static/app/views/releases/thresholdsList/thresholdGroupRows.tsx (hunk 9 above)
Comment (nhsiehgit) [reply static/app/views/releases/thresholdsList/thresholdGroupRows.tsx:225]: So - i _was_ thinking of implementing a "save all" button
but i just havent gotten around to that yet 😬

File: static/app/views/releases/thresholdsList/thresholdGroupRows.tsx (hunk 8 above)
Comment (nhsiehgit) [reply static/app/views/releases/thresholdsList/thresholdGroupRows.tsx:142]: Ah nice.
👍

File: static/app/utils/formatters.tsx [hunk 11]
```diff
@@ -269,6 +269,45 @@ export function getExactDuration(
...
+/**
+ * Given a length of time in seconds, provide me the largest divisible suffix and value for that time period.
+ * eg. 60 -> [1, 'minutes']
+ * eg. 7200 -> [2, 'hours']
+ * eg. 7260 -> [121, 'minutes']
+ *
+ * @param seconds
+ * @param maxSuffix     determines the largest suffix we should pin the response to
+ */
+export function parseLargestSuffix(
```
Comment (sentaur-athena) [inline static/app/utils/formatters.tsx:289]: Since this is a core util method now it can be touched by many people. can you write a short test for it?

File: static/app/views/releases/utils/types.tsx [hunk 12]
```diff
@@ -6,12 +6,12 @@ export type ThresholdQuery = {
 };
 
 export type Threshold = {
-  date_added: string;
   environment: Environment;
   id: string;
   project: Project;
   threshold_type: string;
-  trigger_type: string;
+  trigger_type: string; // 'over' or 'under'
```
Comment (sentaur-athena) [inline static/app/views/releases/utils/types.tsx:13]: Enum?

File: static/app/utils/formatters.tsx (hunk 11 above)
Comment (sentaur-athena) [reply static/app/utils/formatters.tsx:289]: Also where is this being used?

Comment (nhsiehgit) [reply static/app/utils/formatters.tsx:289]: Utilized in the `thresholdGroupRows`

Comment (sentaur-athena) [reply static/app/utils/formatters.tsx:289]: Oops, it was folded when I searched 🤦‍♀️

File: static/app/views/releases/thresholdsList/thresholdGroupRows.tsx [hunk 13]
```diff
@@ -0,0 +1,417 @@
...
+      hasError: false,
+    };
+    setEditingThresholds(updatedEditingThresholds);
+  };
+
+  const saveThreshold = (thresholdIds: string[]) => {
+    thresholdIds.forEach(id => {
+      const thresholdData = editingThresholds[id];
+      const seconds = moment
+        // @ts-ignore: ts is unhappy with moment string duration format
```
Comment (sentaur-athena) [inline static/app/views/releases/thresholdsList/thresholdGroupRows.tsx:112]: What's the warning? Can we address it without ignoring?

Comment (nhsiehgit) [reply static/app/views/releases/thresholdsList/thresholdGroupRows.tsx:112]: hmmm. 
good call. I'll just type it all out to match moments expected types

**Review by sentaur-athena — APPROVED** [review]
